### ISSUE TITLE
[Feature request] Context-free navigation

### ISSUE THREAD

**Issue by KasMA1990** [pop-os/cosmic-epoch#759]
After using Windows for many years, macOS for a few years, and occasionally dabbling in Ubuntu, I've come to realize that the way I prefer to navigate around on my computer is really simple: when I think of an application I need to switch to, I want to see that application immediately.

In other words, I want to see that application immediately, regardless of:
- When was I last using it
- On what virtual desktop I was last using it
- How many windows it has open
- and so on

All of the above are pieces of context I need to keep in mind when navigating around on my computer. Alt-Tab switches between apps based on recency, selecting an open window of the application requires visually parsing my windows, etc.

Windows a really good solution for this: for any applications open in the task bar, you can jump between them by hitting the Windows+number key (with the number matching the index of the application on the bar). With a few pinned applications, I _know_ that I can always get to my browser with Windows+1, terminal with Windows+2, IDE with Windows+3, etc. and I never have to care what's been going on in the system since I used them last. I'm essentially accessing the application I want in constant time, which is lovely for me at least.

On macOS, there's software like [rcmd](https://lowtechguys.com/rcmd/), which also exists to solve this kind of problem. It essentially re-assigns the right Cmd button on the keyboard to always be for jumping between applications, and you use the first letter of the name of application to select what jump to (e.g. RCmd+F opens Files, etc.)

I don't know what the exact solution should look like for COSMIC, but I really hope it's something that can be baked into COSMIC still. There's plenty of discussion around how good design affords people a sense of place in their desktop, but I think many of the concerns there can be sidestepped in cases where users just know where they want to arrive, but don't care about where they are or the path to get there (in terms of navigating the UI).

I know some of this workflow can be approximated with virtual desktops, but I never took a liking to them. To me, the act of organizing applications into different groups just becomes overhead. It's like adding a layer of abstraction to my desktop that I don't need. I just need to jump between applications, and adding layers to navigate only increases the mental burden of things I need to keep in mind.

I'm not trying to find a new paradigm to replace existing navigation with either, so I don't want to take anything away from others who have a good time working with the existing tooling. But I do hope I can find something that better suits myself at least, and I suspect many other power users would start to use something like this if it became available 😊

**Comment by jacobgkau** [comment]
Searching for an app/window in the Launcher will generally let you switch to or launch the app whether it's open already or not, without having to know where it is or how recently it was last activated. It does require a few keystrokes and can be affected by the titles of the open windows, though.

**Comment by KasMA1990** [comment]
Yeah, and I appreciate search helping here, though it still is a bit context dependant. Both as you say, but also:
- I have to parse the search results to see what I'm selecting 
- search results may change for "hidden" reasons

I don't know how the searching in COSMIC works exactly, so I may be wrong here, but when search tries to be clever, it gets harder to predict for users.
For example, searching for "fi" may return "Files" and "Films" (in that order), but then searching for "fil" returns "Films" first and "Files" second.
Or if search results are based on keywords that aren't readily apparent, or fuzzy matching gives weird results.

Experience tells me that although searching can do it, it also requires too much mental effort that I really want to base my workflow on it.

And even if search was perfect, I'd still rather have something like a keyboard shortcut that is as fast and predictable as it can get 😊

**Comment by nevdelap** [comment]
In i3 I basically have a workspace per application or per major use of an application and have function keys F1 thorough F12 bound to workspaces 1 through 12. So for example Chrome with work email, calendar, JIRA is on workspace 1, Slack is on workspace 2, Zoom is on workspace 3, Chrome with personal email, WhatsApp etc is on workspace 4, VS Code is on workspace 5, etc, etc. That is enabled by rules that open things on the right workspace and means I can get to each thing with a single keystroke. 

Cosmic epoch doesn't support bindings to workspaces or the rules to control on which workspace things open yet, but there are issues open and messages saying they are in the works, so in the meantime, since I've moved to Cosmic full time for WFM, I'm just putting things on workspaces manually and putting to with Super+left/right for now.

**Comment by KasMA1990** [comment]
I can definitely see how those features can solve the need, but I'm also curious: are people requesting these features for workspaces because they're lacking better ways to do context free navigation?

Not that those features shouldn't be there, but maybe they're also signs of a different underlying need?

**Comment by brandonbeaumia** [comment]
I am also strongly invested in this paradigm, as are others.

https://github.com/mkropat/jumpapp
https://apps.apple.com/us/app/rcmd-app-switcher/id1596283165?mt=12
https://github.com/Soft/run-or-raise
https://unix.stackexchange.com/questions/656729/wmctrl-like-tool-or-alternative-for-kde-wayland
https://github.com/adrienverge/gnome-magic-window

Unfortunately, every solution I have found is either a GNOME extension or incompatible with Wayland. Having a feature that lets me set a keybinding for an application that

1) launches a specified app if it isn't running, 
2) focuses the app if it is, and 
3) cycles through open windows _of_ that app if it's already focused

would be incredibly efficient. I hacked together Kanata and AutoHotkey to do this at on Windows at work and this "context-free" navigation is a really great experience; I find myself missing it so much on Cosmic.

**Comment by nevdelap** [comment]
**Update** Being able to bind function keys to workspaces is supported. I don't know when that become the case, but I discovered it yesterday. You do have to manually edit the config because cosmic-settings either doesn't support it yet, or the support has a bug. I opened this yesterday: pop-os/cosmic-settings#597

There is one problem though, because each display has workspaces numbered from 1, rather being able assign workspaces to displays, it works in the context of the display that has focus.

**Comment by KasMA1990** [comment]
@brandonbeaumia in your own experience, what kind of shortcuts have you preferred having? As in, do you just pick random unused shortcuts, or have you had a systematic approach to them?

**Comment by brandonbeaumia** [comment]
@KasMA1990 I use Kanata to reprogram my tab key to "tab when tapped, Ctrl+Alt+Shift when held," then bind that combo with letters by mnemonic (tab+F for files, tab+S for spreadsheets, etc). I also try very hard to put the bindings on the left-hand side only so I don't have to leave the mouse.

**Comment by KasMA1990** [comment]
Using tab as a modifier is a pretty cool idea 😄

**Comment by petrem** [comment]
The one workspace per app or activity worked for me well a long long time. The problem with this approach, in Cosmic, is that the workspaces appear and disappear dynamically. Say I open my editor in workspace 1 and the browser in workspace2 and so on. If I restart my editor, suddenly everything rotates to the left with the browser being in the first workspace. There goes the context-free-ness of it.
I'd love to have the static workspaces back, please (and also I'd like the option for focus and moving of the windows to not automatically cross workspaces).

**Comment by pieter-groenendijk** [comment]
> The one workspace per app or activity worked for me well a long long time. The problem with this approach, in Cosmic, is that the workspaces appear and disappear dynamically. Say I open my editor in workspace 1 and the browser in workspace2 and so on. If I restart my editor, suddenly everything rotates to the left with the browser being in the first workspace. There goes the context-free-ness of it. I'd love to have the static workspaces back, please (and also I'd like the option for focus and moving of the windows to not automatically cross workspaces).

I agree with @petrem, I've always used static workspaces with static usecases. This way I know where to find my running applications without having to look for them, i.e. I can reach them with absolute specifiers: workspace 1, 2 or `n`.

**In my head it comes down to creating a workspace when a workspace position is asked for but not yet in use, no matter if the workspace specifier is relative or absolute. The dynamic collapsing behavior of workspaces would need some thought. I feel like this can be a non-intrusive and consistent change.** 

Thank you for reading, and good luck on this great project :). I'm excited.

**Comment by lapause** [comment]
> I am also strongly invested in this paradigm, as are others.
> 
> https://github.com/mkropat/jumpapp https://apps.apple.com/us/app/rcmd-app-switcher/id1596283165?mt=12 https://github.com/Soft/run-or-raise https://unix.stackexchange.com/questions/656729/wmctrl-like-tool-or-alternative-for-kde-wayland https://github.com/adrienverge/gnome-magic-window
> 
> Unfortunately, every solution I have found is either a GNOME extension or incompatible with Wayland. Having a feature that lets me set a keybinding for an application that
> 
>     1. launches a specified app if it isn't running,
> 
>     2. focuses the app if it is, and
> 
>     3. cycles through open windows _of_ that app if it's already focused
> 
> 
> would be incredibly efficient. I hacked together Kanata and AutoHotkey to get custom keybinds for this on Windows at work (rather than Win+num) and this "context-free" navigation is a really great experience; I find myself missing it so much on Cosmic.

I've got a good news for you, I was having the same issues and just made a tool that does that (and more): https://github.com/lapause/cosmic-ext-window-helper
